File: frontend/src/lib/api/transactions.ts
import { api } from "@/api/client";
import type { Transaction, TransactionCategory } from "@/types";

export type RecurringFrequency = "weekly" | "biweekly" | "monthly";

export interface RecurringTransaction {
  id: string;
  name: string;
  amount: number;
  currency: string;
  transaction_type: "expense" | "income";
  frequency: RecurringFrequency;
  day_of_month: number;
  reminder_days_before: number;
  is_active: boolean;
  start_date: string;
  end_date?: string;
}

export interface RecurringTransactionCreate {
  account_id: string;
  category_id?: string;
  name: string;
  amount: number;
  currency?: string;
  transaction_type: "expense" | "income";
  frequency: RecurringFrequency;
  day_of_month?: number;
  start_date: string;
  end_date?: string;
  description?: string;
}

export interface TransactionFilters {
  account_id?: string;
  category_id?: string;
  start_date?: string;
  end_date?: string;
  limit?: number;
  offset?: number;
}

export interface TransactionCreate {
  account_id: string;
  category_id?: string;
  transaction_type: "expense" | "income" | "transfer";
  amount: number;
  currency?: string;
  transaction_date: string;
  description?: string;
  merchant?: string;
}

export const transactionsApi = {
  list: async (filters?: TransactionFilters): Promise<Transaction[]> =>
    (await api.get("/v1/transactions/", { params: filters })).data,

  create: async (data: TransactionCreate): Promise<Transaction> =>
    (await api.post("/v1/transactions/", data)).data,

  update: async (id: string, data: Record<string, unknown>): Promise<Transaction> =>
    (await api.patch(`/v1/transactions/${id}`, data)).data,

  delete: async (id: string): Promise<void> => {
    await api.delete(`/v1/transactions/${id}`);
  },

  listCategories: async (): Promise<TransactionCategory[]> =>
    (await api.get("/v1/transactions/categories")).data,

  // ── Transakcje cykliczne ──
  listRecurring: async (): Promise<RecurringTransaction[]> =>
    (await api.get("/v1/transactions/recurring/list")).data,

  createRecurring: async (data: RecurringTransactionCreate): Promise<RecurringTransaction> =>
    (await api.post("/v1/transactions/recurring", data)).data,

  deactivateRecurring: async (id: string): Promise<void> => {
    await api.delete(`/v1/transactions/recurring/${id}`);
  },
};

File: backend/app/api/v1/endpoints/transactions.py
import uuid
from datetime import date
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.base import get_db
from app.models.user import User
from app.schemas.transaction import (
    RecurringTransactionCreate,
    RecurringTransactionResponse,
    TagCreate,
    TagResponse,
    TransactionCategoryCreate,
    TransactionCategoryResponse,
    TransactionCreate,
    TransactionResponse,
    TransactionUpdate,
)
from app.services.transaction import TransactionService

router = APIRouter(prefix="/transactions", tags=["Transactions"])


@router.get("/categories", response_model=list[TransactionCategoryResponse])
async def list_categories(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.list_categories(current_user.id)


@router.post("/categories", response_model=TransactionCategoryResponse, status_code=201)
async def create_category(
    payload: TransactionCategoryCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.create_category(
        current_user.id, payload.name, payload.parent_id, payload.icon, payload.color
    )


@router.get("/tags", response_model=list[TagResponse])
async def list_tags(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.list_tags(current_user.id)


@router.post("/tags", response_model=TagResponse, status_code=201)
async def create_tag(
    payload: TagCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.create_tag(current_user.id, payload.name)


@router.post("/", response_model=TransactionResponse, status_code=201)
async def create_transaction(
    payload: TransactionCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.create(current_user.id, **payload.model_dump())


@router.get("/", response_model=list[TransactionResponse])
async def list_transactions(
    account_id: Optional[uuid.UUID] = None,
    category_id: Optional[uuid.UUID] = None,
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    limit: int = Query(default=50, le=500),
    offset: int = 0,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.list(
        current_user.id,
        account_id=account_id,
        category_id=category_id,
        start_date=start_date,
        end_date=end_date,
        limit=limit,
        offset=offset,
    )


@router.get("/{transaction_id}", response_model=TransactionResponse)
async def get_transaction(
    transaction_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.get_owned(transaction_id, current_user.id)


@router.patch("/{transaction_id}", response_model=TransactionResponse)
async def update_transaction(
    transaction_id: uuid.UUID,
    payload: TransactionUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    tx = await svc.get_owned(transaction_id, current_user.id)
    return await svc.update(tx, **payload.model_dump(exclude_none=True))


@router.delete("/{transaction_id}", status_code=204)
async def delete_transaction(
    transaction_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    tx = await svc.get_owned(transaction_id, current_user.id)
    await svc.delete(tx)


@router.post("/recurring", response_model=RecurringTransactionResponse, status_code=201)
async def create_recurring(
    payload: RecurringTransactionCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.create_recurring(user_id=current_user.id, **payload.model_dump())


@router.get("/recurring/list", response_model=list[RecurringTransactionResponse])
async def list_recurring(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    return await svc.list_active_recurring(current_user.id)


@router.delete("/recurring/{recurring_id}", status_code=204)
async def deactivate_recurring(
    recurring_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TransactionService(db)
    await svc.deactivate_recurring(recurring_id, current_user.id)
